File: check_braces.dart
import 'dart:io';

void main(List<String> args) {
  final file = File(args[0]);
  final content = file.readAsStringSync();
  final lines = content.split('\n');
  
  final stack = <Map<String, dynamic>>[];
  
  for (int i = 0; i < lines.length; i++) {
    final line = lines[i];
    bool inString = false;
    for (int j = 0; j < line.length; j++) {
      final char = line[j];
      if (char == "'" || char == '"') {
        inString = !inString;
      }
      if (inString) continue;
      
      if (char == '{' || char == '[' || char == '(') {
        stack.add({'char': char, 'line': i + 1, 'col': j + 1});
      } else if (char == '}' || char == ']' || char == ')') {
        if (stack.isEmpty) {
          print('Unmatched ' + char + ' at line ' + (i + 1).toString());
          return;
        }
        final top = stack.removeLast();
        if (i + 1 >= 1765 && i + 1 <= 1773) {
           print('Line ' + (i + 1).toString() + ' (' + char + ') closed ' + top['char'].toString() + ' at line ' + top['line'].toString());
        }
        
        final topChar = top['char'];
        String expected = '';
        if (topChar == '{') expected = '}';
        if (topChar == '[') expected = ']';
        if (topChar == '(') expected = ')';
        
        if (char != expected) {
          print('Mismatched ' + char + ' at line ' + (i + 1).toString() + ', expected ' + expected + ' to match ' + topChar + ' at line ' + top["line"].toString());
          return;
        }
      }
    }
  }
}

File: check_braces.py
import sys

def check_braces(filepath):
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    
    stack = []
    lines = content.split('\n')
    for i, line in enumerate(lines):
        for j, char in enumerate(line):
            if char in '{[(':
                stack.append((char, i+1, j+1))
            elif char in '}])':
                if not stack:
                    print(f"Unmatched {char} at line {i+1}")
                    return
                top, line_num, col_num = stack.pop()
                expected = {'{': '}', '[': ']', '(': ')'}[top]
                if char != expected:
                    print(f"Mismatched {char} at line {i+1}, expected {expected} to match {top} at line {line_num}")
                    return
    if stack:
        print("Unclosed braces:")
        for char, line_num, col_num in stack:
            print(f"  {char} at line {line_num}:{col_num}")
    else:
        print("All braces match perfectly.")

check_braces(sys.argv[1])
